File: Haskell/Lexer/Token.hs
module Token where

import Data.Map qualified as M

data TokenType
  = EOF
  | Whitespace
  | Illegal
  | Let
  | Const
  | Struct
  | Function
  | If
  | Else
  | Switch
  | Case
  | Break
  | Return
  | While
  | For
  | And
  | Or
  | In
  | Identifier
  | Number
  | Float
  | Boolean
  | String
  | Char
  | Int
  | F32
  | Bool
  | Str
  | Nil
  | Void
  | Exclamation
  | At
  | Hashtag
  | Dollar
  | Percent
  | Caret
  | Ampersand
  | Asterisk
  | Lparen
  | Rparen
  | Minus
  | Underscore
  | Plus
  | Assign
  | Lbracket
  | Rbracket
  | Lbrace
  | Rbrace
  | Semicolon
  | Colon
  | Apostrophe
  | Quote
  | Comma
  | Period
  | Lessthan
  | Greaterthan
  | Slash
  | Question
  | Backslash
  | Pipe
  | Equal
  | NotEqual
  | PlusEq
  | MinusEq
  | MultEq
  | DivEq
  | LtEqual
  | GtEqual
  | Increment
  | Decrement
  | Comment
  deriving (Enum, Show, Eq)

data Token = Token TokenType String deriving (Show)

type KeywordMap = M.Map String TokenType

keywordMap :: KeywordMap
keywordMap =
  M.fromList
    [ ("let", Let),
      ("const", Const),
      ("struct", Struct),
      ("fn", Function),
      ("if", If),
      ("else", Else),
      ("switch", Switch),
      ("case", Case),
      ("break", Break),
      ("return", Return),
      ("while", While),
      ("for", For),
      ("and", And),
      ("or", Or),
      ("in", In),
      ("true", Boolean),
      ("false", Boolean),
      ("bool", Bool),
      ("int", Int),
      ("f32", F32),
      ("str", Str),
      ("nil", Nil),
      ("void", Void)
    ]

File: Clojure/Lexer/src/lexer/token.clj
(ns lexer.token)

(defrecord Token [type literal])

(def singleTokenMap {\! :exclamation
                     \@ :at
                     \# :hashtag
                     \$ :dollar
                     \% :percent
                     \^ :caret
                     \& :ampersand
                     \* :asterisk
                     \( :lparen
                     \) :rparen
                     \- :minus
                     \_ :underscore
                     \+ :plus
                     \= :assign
                     \[ :lbracket
                     \] :rbracket
                     \{ :lbrace
                     \} :rbrace
                     \; :semicolon
                     \: :colon
                     \' :char
                     \" :string
                     \, :comma
                     \. :period
                     \< :lessthan
                     \> :greaterthan
                     \/ :slash
                     \? :question
                     \\ :backslash
                     \| :pipe})

(def doubleTokenMap {"==" :equal
                     "!=" :not_equal
                     "+=" :plus_eq
                     "-=" :minus_eq
                     "*=" :mult_eq
                     "/=" :div_eq
                     "<=" :lt_equal
                     ">=" :gt_equal
                     "++" :increment
                     "--" :decrement
                     "//" :comment})

(def keywordMap {"let" :let
                 "const" :const
                 "struct" :struct
                 "fn" :function
                 "if" :if
                 "else" :else
                 "switch" :switch
                 "case" :case
                 "break" :break
                 "return" :return
                 "while" :while
                 "for" :for
                 "and" :and
                 "or" :or
                 "in" :in
                 "true" :boolean
                 "false" :boolean
                 "bool" :bool
                 "int" :int
                 "f32" :f32
                 "str" :str
                 "nil" :nil
                 "void" :void})
